seqr/views/utils/test_utils.py

```diff
@@ -91,6 +92,131 @@ def _check_login(self, url, permission_level, request_data=None):
+    wss = list(wss)
+    return wss[0]['acl'] if wss else {}
+
+
+def get_workspaces_side_effect(user, fields):
+    return [
+        {
+            'public': ws['public'],
+            'accessLevel': ws['acl'][user.email]['accessLevel'],
+            'workspace':{
+                'namespace': ws['workspace_namespace'],
+                'name': ws['workspace_name']
+            }
+        } for ws in ANVIL_WORKSPACES if user.email in ws['acl'].keys()
+    ]
+
+
+class AnvilAuthenticationTestCase(AuthenticationTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.staff_user = User.objects.get(username='test_user')
+        cls.manager_user = User.objects.get(username='test_user_manager')
+        cls.collaborator_user = User.objects.get(username='test_user_non_staff')
+        cls.no_access_user = User.objects.get(username='test_user_no_access')
+
+    # mock the terra apis
+    def setUp(self):
+        patcher = mock.patch('seqr.views.utils.terra_api_utils.TERRA_API_ROOT_URL', TEST_TERRA_API_ROOT_URL)
+        patcher.start()
+        self.addCleanup(patcher.stop)
+        patcher = mock.patch('seqr.views.utils.terra_api_utils.time')
+        patcher.start().return_value = TOKEN_AUTH_TIME + 10
+        self.addCleanup(patcher.stop)
+        patcher = mock.patch('seqr.views.utils.permissions_utils.list_anvil_workspaces')
+        self.mock_list_workspaces = patcher.start()
+        self.mock_list_workspaces.side_effect = get_workspaces_side_effect
+        self.addCleanup(patcher.stop)
+        patcher = mock.patch('seqr.views.utils.terra_api_utils._service_account_session')
+        self.mock_service_account = patcher.start()
+        self.mock_service_account.get.side_effect = get_ws_acl_side_effect
+        self.addCleanup(patcher.stop)
+
+
+# the workspace list returned by mocked terra api. It makes the results of the mixed tests unique to local and anvil tests
+MIX_TEST_WORKSPACES = [
+        {
+            'public': False,
+            'accessLevel': 'READER',
+            'workspace': {
+                'namespace': 'my-seqr-billing',
+                'name': 'anvil-project 1000 Genomes Demo'
+            }
+        }, {
+           'public': False,
+           'accessLevel': 'OWNER',
+           'workspace': {
+               'namespace': 'my-seqr-billing',
+               'name': 'anvil-no-project-workspace1'
+           }
+        }, {
+           'public': True,
+           'accessLevel': 'READER',
+           'workspace': {
+               'namespace': 'my-seqr-billing',
+               'name': 'anvil-no-project-workspace2'
+           }
+        },
+    ]
+
+
+class MixAuthenticationTestCase(AnvilAuthenticationTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        AuthenticationTestCase.setUpTestData()
+
+    def setUp(self):
+        super(MixAuthenticationTestCase, self).setUp()
+        self.mock_list_workspaces.side_effect = lambda user, fields: MIX_TEST_WORKSPACES\
```

Comment:
Is there a reason this needs to be different? What edge case does this test that you can't test by using a different user with the same AnVIL permissions? Would adding a test for the manager user allow you to cover that edge case while using the same permissions for everything? Or maybe actually inspecting some of the json itself (all the projects json includes whether or not the user has edit permissions, so if those differ between anvil and local and mix maybe that would be a way to test it)? Ideally, the mixed test should be different because the local and anvil permissions ARE different, and the mixed permissions are therefore different than both of them

If this is really necessary to have different mocked data in anvil, can you add a comment explaining why?

Reply:
The different mocked data is not always necessary especially when the local projects have left some permissions which can be filled by the anvil permissions. Base on current local fixture permissions, I probably can do as below table:
|                | local |         |              | Anvil |         |              | Mix |         |              |
|----------------|-------|---------|--------------|-------|---------|--------------|-------------|---------|--------------|
| project No.    | staff | manager | collaborator | staff | manager | collaborator | staff       | manager | collaborator |
| 1              | y     | y       | y            | na    | y      | na           | y           | y       | y            |
| 2              | y     | y       | y            | na    | na      | na           | y           | y       | y            |
| 3              | y     | y       | n            | na    | y      | y            | y           | y       | n            |
| total projects | 3     | 3       | `2`            | 3     | 2       | `1 `           | 3           | 3       | `3`            |  

The numbers of projects for the collaborator are different for local/anvil/mix test cases.